Data/ByteString.hs

```diff
@@ -1148,14 +1148,12 @@ elemIndexEnd = findIndexEnd . (==)
 elemIndices :: Word8 -> ByteString -> [Int]
 elemIndices w (BS x l) = loop 0
     where
-        loop !n = let q = accursedUnutterablePerformIO $ withForeignPtr x $ \p ->
-                           memchr (p `plusPtr` n)
-                                                w (fromIntegral (l - n))
-                  in if q == nullPtr
-                        then []
-                        else let i = accursedUnutterablePerformIO $ withForeignPtr x $ \p ->
-                                       return (q `minusPtr` p)
-                             in i : loop (i+1)
+        loop !n = accursedUnutterablePerformIO $ withForeignPtr x $ \p -> do
+            q <- memchr (p `plusPtr` n) w (fromIntegral (l - n))
+            if q == nullPtr
+                then return []
+                else let !i = q `minusPtr` p
+                      in return $! i : loop (i + 1)
```

Comment:
What's the reason for using `$!` instead of `$` here? The `(:)` constructor application is in WHNF anyway, no?!

```suggestion
                      in return $ i : loop (i + 1)
```

Reply:
Mostly habit, when I am not 100% sure how strictness propagates through expressions like `return $ Foo $ a`, where strictness in the result need not force `a`.  Now it is true that here I also have `let !i = ...` so the expression `return $ (:) (i) (loop (i+i))` should be strict in `i` even without `return $!`, but my impression is that the extra annotation is "free" (no additional runtime cost), and is then just a matter of expression of intent than necessity.

Perhaps what I really mean is:   `return $ (:) $! i $ loop (i+1)`, that is don't hide an unevaluated `i` inside the list constructor, so you're perhaps right that the `return $!` may not do what I hope it does.  I just want to ensure that the first list element is evaluated strictly.  Is it important to change this in your view?

Reply:
> Is it important to change this in your view?

I'm slightly curious whether the redundant forcing is optimized away. I don't expect it to matter much or at all though.

Reply:
@hsyl20 probably knows the answer off hand. Any advice?

Reply:
Yes `$!` isn't useful here and it should be optimized away. It could be clearer to replace it with `$` to avoid people thinking that the whole list is forced or that the intent was to force the whole list and that it's a bug.